File: api/config_migrations.py
"""config.json のスキーマ・マイグレーション（純粋な dict 変換）。

ファイル I/O やロックは持たず、config dict を受け取って変換後の dict を返す
純粋関数のみを置く。config.py（I/O 層）から呼び出される一方向依存とする。

マイグレーションはスキーマバージョンの段階適用（__global__.config_version を基準）
のみ。破壊的なスキーマ変更は `_CONFIG_MIGRATIONS` に変換関数を登録して行う。
"""

import logging
from collections.abc import Callable

from .common import (
    CONFIG_SCHEMA_VERSION,
    GLOBAL_CONFIG_KEY,
)

logger = logging.getLogger(__name__)


def _get_config_version(config: dict) -> int:
    """config に保存されたスキーマバージョンを返す。未設定/不正なら 0（旧版）。"""
    global_section = config.get(GLOBAL_CONFIG_KEY)
    if isinstance(global_section, dict):
        version = global_section.get("config_version")
        if isinstance(version, int) and not isinstance(version, bool) and version >= 0:
            return version
    return 0


def _set_config_version(config: dict, version: int) -> dict:
    """__global__.config_version を書き込んだ新しい config を返す。"""
    global_section = config.get(GLOBAL_CONFIG_KEY)
    global_section = dict(global_section) if isinstance(global_section, dict) else {}
    global_section["config_version"] = version
    new_config = dict(config)
    new_config[GLOBAL_CONFIG_KEY] = global_section
    return new_config


# 旧バージョン -> 次バージョンへの変換関数。キー N の関数は version N の config を
# version N+1 に変換する。破壊的なスキーマ変更を入れる際にここへ追加する。
# 各関数は config dict を受け取り、変換後の dict を返す（元を破壊しないこと）。
_CONFIG_MIGRATIONS: dict[int, Callable[[dict], dict]] = {}


def _migrate_config_version(config: dict) -> tuple[dict, bool]:
    """config を CONFIG_SCHEMA_VERSION まで段階的にマイグレーションする。

    - 空 config（初回起動）は何もしない。
    - 旧版はマイグレーションを順次適用し、バージョンを刻んで返す。
    - コードより新しいバージョンの config は破壊を避けるため変換も再書き込みも
      せず、警告のみ出してそのまま返す（best-effort 互換動作）。
    """
    if not config:
        return config, False

    current = _get_config_version(config)
    if current > CONFIG_SCHEMA_VERSION:
        logger.warning(
            "config_version %d is newer than supported %d; "
            "running in best-effort compatibility mode",
            current, CONFIG_SCHEMA_VERSION,
        )
        return config, False
    if current == CONFIG_SCHEMA_VERSION:
        return config, False

    migrated = config
    version = current
    while version < CONFIG_SCHEMA_VERSION:
        migrate = _CONFIG_MIGRATIONS.get(version)
        if migrate is not None:
            migrated = migrate(migrated)
        version += 1
    migrated = _set_config_version(migrated, CONFIG_SCHEMA_VERSION)
    logger.info("migrated config schema v%d -> v%d", current, CONFIG_SCHEMA_VERSION)
    return migrated, True

File: server/src/config_migrations.rs
//! config.json のスキーマバージョン管理。
//!
//! ファイル I/O やロックは持たず、config dict を受け取って変換後の dict を返す
//! 純粋関数のみを置く。

use serde_json::{Map, Value};

use crate::config::GLOBAL_CONFIG_KEY;

pub const CONFIG_SCHEMA_VERSION: i64 = 4;

/// config に保存されたスキーマバージョンを返す。未設定/不正なら 0（旧版）。
pub fn get_config_version(config: &Map<String, Value>) -> i64 {
    config
        .get(GLOBAL_CONFIG_KEY)
        .and_then(Value::as_object)
        .and_then(|g| g.get("config_version"))
        .and_then(|v| if v.is_boolean() { None } else { v.as_i64() })
        .filter(|&v| v >= 0)
        .unwrap_or(0)
}

fn set_config_version(mut config: Map<String, Value>, version: i64) -> Map<String, Value> {
    let mut global = config
        .get(GLOBAL_CONFIG_KEY)
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    global.insert("config_version".to_string(), Value::Number(version.into()));
    config.insert(GLOBAL_CONFIG_KEY.to_string(), Value::Object(global));
    config
}

/// config の config_version を CONFIG_SCHEMA_VERSION に揃える。
/// v4 未満からのキー変換（radial / pinned_jobs / detached_tab）は打ち切ったため、
/// 旧バージョンの config はバージョンだけ付け直し、旧キーの設定は引き継がない。
/// 戻り値の bool は書き戻しが必要かどうか。
pub fn migrate_config_version(config: Map<String, Value>) -> (Map<String, Value>, bool) {
    if config.is_empty() {
        return (config, false);
    }
    let current = get_config_version(&config);
    if current > CONFIG_SCHEMA_VERSION {
        tracing::warn!(
            "config_version {} is newer than supported {}; running in best-effort compatibility mode",
            current,
            CONFIG_SCHEMA_VERSION
        );
        return (config, false);
    }
    if current == CONFIG_SCHEMA_VERSION {
        return (config, false);
    }
    if current > 0 {
        tracing::warn!(
            "config_version {} is no longer migrated; settings stored under pre-v{} keys are ignored",
            current,
            CONFIG_SCHEMA_VERSION
        );
    }
    (set_config_version(config, CONFIG_SCHEMA_VERSION), true)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn as_map(v: Value) -> Map<String, Value> {
        v.as_object().unwrap().clone()
    }

    #[test]
    fn version_detection() {
        assert_eq!(get_config_version(&Map::new()), 0);
        assert_eq!(
            get_config_version(&as_map(json!({"__global__": {"config_version": 3}}))),
            3
        );
        assert_eq!(
            get_config_version(&as_map(json!({"__global__": {"config_version": true}}))),
            0
        );
        assert_eq!(
            get_config_version(&as_map(json!({"__global__": {"config_version": -1}}))),
            0
        );
    }

    #[test]
    fn empty_config_untouched() {
        let (out, migrated) = migrate_config_version(Map::new());
        assert!(out.is_empty());
        assert!(!migrated);
    }

    #[test]
    fn old_version_is_stamped_without_converting_legacy_keys() {
        let cfg = as_map(json!({
            "__global__": {"config_version": 2, "pinned_jobs": [{"key": "k1"}]},
        }));
        let (out, migrated) = migrate_config_version(cfg);
        assert!(migrated);
        let global = out["__global__"].as_object().unwrap();
        assert_eq!(global["config_version"], json!(CONFIG_SCHEMA_VERSION));
        assert!(!global.contains_key("recent_jobs"));
    }

    #[test]
    fn newer_version_left_untouched() {
        let cfg = as_map(json!({"__global__": {"config_version": 99, "future_field": 1}}));
        let (out, migrated) = migrate_config_version(cfg.clone());
        assert!(!migrated);
        assert_eq!(out, cfg);
    }

    #[test]
    fn current_version_no_write() {
        let cfg = as_map(json!({"__global__": {"config_version": 4}}));
        let (_, migrated) = migrate_config_version(cfg);
        assert!(!migrated);
    }
}
